File: language-model/gpu/vram_calculator.py
import os
import json
import urllib.request
import argparse

# Calculate GPU VRAM requirements for LLM models
# 
# Version 1.5

GGUF_MAPPING = {
    "Q8_0": 8.5,
    "Q6_K": 6.59,
    "Q5_K_M": 5.69,
    "Q5_K_S": 5.54,
    "Q5_0": 5.54,
    "Q4_K_M": 4.85,
    "Q4_K_S": 4.58,
    "Q4_0": 4.55,
    "IQ4_NL": 4.5,
    "Q3_K_L": 4.27,
    "IQ4_XS": 4.25,
    "Q3_K_M": 3.91,
    "IQ3_M": 3.7,
    "IQ3_S": 3.5,
    "Q3_K_S": 3.5,
    "Q2_K": 3.35,
    "IQ3_XS": 3.3,
    "IQ3_XXS": 3.06,
    "IQ2_M": 2.7,
    "IQ2_S": 2.5,
    "IQ2_XS": 2.31,
    "IQ2_XXS": 2.06,
    "IQ1_S": 1.56,
}

EXL2_OPTIONS = [round(i, 2) for i in reversed([x * 0.05 + 2 for x in range(81)])]

def bits_to_gb(bits):
    return bits / (2 ** 30)


def calculate_vram_raw(num_params=None, bpw=5.0, lm_head_bpw=6.0, kv_cache_bpw=8.0, context=None, fp8=True,
                       num_gpus=1, max_position_embeddings=None, num_hidden_layers=None, hidden_size=None,
                       num_key_value_heads=None, num_attention_heads=None, intermediate_size=None, vocab_size=None,
                       gqa=True):
    cuda_size = 500 * 2 ** 20 * num_gpus
    params_size = num_params * 1e9 * (bpw / 8)
    kv_cache_size = (context * 2 * num_hidden_layers * hidden_size) * (kv_cache_bpw / 8)

    if gqa:
        kv_cache_size *= num_key_value_heads / num_attention_heads

    bytes_per_param = (bpw / 8)
    lm_head_bytes_per_param = (lm_head_bpw / 8)
    head_dim = hidden_size / num_attention_heads

    attention_input = bytes_per_param * context * hidden_size
    q = bytes_per_param * context * head_dim * num_attention_heads
    k = bytes_per_param * context * head_dim * num_key_value_heads
    v = bytes_per_param * context * head_dim * num_key_value_heads

    softmax_output = lm_head_bytes_per_param * num_attention_heads * context
    softmax_dropout_mask = num_attention_heads * context
    dropout_output = lm_head_bytes_per_param * num_attention_heads * context
    out_proj_input = lm_head_bytes_per_param * context * num_attention_heads * head_dim
    attention_dropout = context * hidden_size

    attention_block = (
        attention_input + q + k + softmax_output + v + out_proj_input + softmax_dropout_mask + dropout_output + attention_dropout
    )

    mlp_input = bytes_per_param * context * hidden_size
    activation_input = bytes_per_param * context * intermediate_size
    down_proj_input = bytes_per_param * context * intermediate_size
    dropout_mask = context * hidden_size
    mlp_block = mlp_input + activation_input + down_proj_input + dropout_mask

    layer_norms = bytes_per_param * context * hidden_size * 2
    activations_size = attention_block + mlp_block + layer_norms
    output_size = lm_head_bytes_per_param * context * vocab_size

    vram_bits = cuda_size + params_size + activations_size + output_size + kv_cache_size
    return round(bits_to_gb(vram_bits), 2)


def download_file(url, filepath, headers=None):
    if headers is None:
        headers = {}

    # Create the cache directory if it doesn't exist
    os.makedirs(os.path.dirname(filepath), exist_ok=True)

    # Check if file already exists
    if os.path.exists(filepath):
        return

    req = urllib.request.Request(url, headers=headers)
    with urllib.request.urlopen(req) as response, open(filepath, 'wb') as out_file:
        while chunk := response.read(1024):
            out_file.write(chunk)


def download_model_configs(model_id, access=None):
    url = f"https://huggingface.co/{model_id}"
    readme_url = f"{url}/raw/main/README.md"
    config_url = f"{url}/raw/main/config.json"
    index_url = f"{url}/raw/main/model.safetensors.index.json"

    cache_dir = os.path.join("cache", model_id)
    readme_path = os.path.join(cache_dir, "README.md")
    config_path = os.path.join(cache_dir, "config.json")
    index_path = os.path.join(cache_dir, "model.safetensors.index.json")

    headers = {}
    if access:
        headers["Authorization"] = f"Bearer {access}"

    download_file(readme_url, readme_path, headers=headers)
    download_file(config_url, config_path, headers=headers)
    download_file(index_url, index_path, headers=headers)


def get_model_config(model_id, access=None):
    download_model_configs(model_id, access=access)
    cache_dir = os.path.join("cache", model_id)
    config_path = os.path.join(cache_dir, "config.json")
    index_path = os.path.join(cache_dir, "model.safetensors.index.json")
    config = json.load(open(config_path))
    index = json.load(open(index_path))
    num_params = index["metadata"]["total_size"] / 2 / 1e9
    config["num_params"] = num_params
    return config


def parse_bpw(bpw):
    try:
        return float(bpw)
    except ValueError:
        bpw = bpw.upper()
        return GGUF_MAPPING.get(bpw, 5.0)  # Default to 5.0 if not found

def get_bpw_values(bpw, fp8):
    bpw = parse_bpw(bpw)
    lm_head_bpw = 8.0 if bpw > 6.0 else 6.0
    kv_cache_bpw = 8 if fp8 else 16
    return {
        "bpw": bpw,
        "lm_head_bpw": lm_head_bpw,
        "kv_cache_bpw": kv_cache_bpw,
    }

def calculate_vram(model_id, bpw=5.0, context=None, fp8=True, access=None):
    config = get_model_config(model_id, access=access)
    bpw_values = get_bpw_values(bpw, fp8)
    context = context or config["max_position_embeddings"]
    return calculate_vram_raw(
        num_params=config["num_params"],
        bpw=bpw_values["bpw"],
        lm_head_bpw=bpw_values["lm_head_bpw"],
        kv_cache_bpw=bpw_values["kv_cache_bpw"],
        context=context,
        num_hidden_layers=config["num_hidden_layers"],
        hidden_size=config["hidden_size"],
        num_key_value_heads=config["num_key_value_heads"],
        num_attention_heads=config["num_attention_heads"],
        intermediate_size=config["intermediate_size"],
        vocab_size=config["vocab_size"],
    )


def calculate_context(model_id, memory=48, bpw=5.0, fp8=True, access=None):
    config = get_model_config(model_id, access=access)
    min_context = 2048
    max_context = config["max_position_embeddings"]

    low, high = min_context, max_context
    while low < high:
        mid = (low + high + 1) // 2
        if calculate_vram(model_id, bpw, mid, fp8, access=access) > memory:
            high = mid - 1
        else:
            low = mid

    context = low
    while calculate_vram(model_id, bpw, context, fp8, access=access) < memory and context <= max_context:
        context += 100

    return context - 100


def calculate_bpw(model_id, memory=48, context=None, fp8=True, type="exl2", access=None):
    if type == "exl2":
        for bpw in EXL2_OPTIONS:
            if calculate_vram(model_id, bpw, context, fp8, access=access) < memory:
                return bpw
    elif type == "gguf":
        for name, bpw in GGUF_MAPPING.items():
            if calculate_vram(model_id, bpw, context, fp8, access=access) < memory:
                return name
    return None


def main(model_id, mode="A", bpw=5.0, context=None, memory=48, type="exl2", fp8=True, access=None):
    if mode == "A":
        vram = calculate_vram(model_id, bpw, context, fp8, access=access)
        print(vram)
    elif mode == "B":
        context = calculate_context(model_id, memory, bpw, fp8, access=access)
        print(context)
    elif mode == "C":
        bpw = calculate_bpw(model_id, memory, context, fp8, type=type, access=access)
        print(bpw)


if __name__ == "__main__":
    parser = argparse.ArgumentParser(description="VRAM, context length, and BPW calculator for LLMs.")
    parser.add_argument("model_id", type=str, help="Model ID from Hugging Face Hub")
    parser.add_argument("-m", "--mode", choices=["A", "B", "C"], default="A", help="Mode: A (calculate VRAM required), B (calculate context you can fit in your VRAM), C (calculate the best BPW you can fit in your VRAM)")
    parser.add_argument("-b", "--bpw", type=str, default="5.0", help="Bits per weight (can be float or GGUF ID, default: 5.0)")
    parser.add_argument("-c", "--context", type=int, default=None, help="Context length (default: use model setting)")
    parser.add_argument("-r", "--ram", type=int, default=48, help="Available VRAM in GB (default: 48)")
    parser.add_argument("-t", "--type", type=str, default="exl2", choices=["exl2", "gguf"],
                        help="Type of quantization [exl2, gguf] for BPW calculation (default: exl2)")
    parser.add_argument("-f", "--fp8", action=argparse.BooleanOptionalAction, help="Use FP8 KV cache (default: false)")
    parser.add_argument("-a", "--access", type=str, default=None, help="Access token for Hugging Face API (optional)")

    args = parser.parse_args()
    main(
        model_id=args.model_id,
        mode=args.mode,
        bpw=args.bpw,
        context=args.context,
        memory=args.ram,
        type=args.type,
        fp8=args.fp8,
        access=args.access
    )


"""
Examples:

1) How much memory do I need to run a LLama 3.1 405B model quantized to Q4_0?
$ python3 vram_calculator.py --mode A --bpw Q4_0 --access hf_xxxxxxxx meta-llama/Meta-Llama-3.1-405B-Instruct-FP8
288.41


2) How much context can I get out of this model?
$ python3 vram_calculator.py --mode B --ram 289 --bpw 4.55 meta-llama/Meta-Llama-3.1-405B-Instruct-FP8
131072


3) What is the best quant I can run of this model?
$ python3 vram_calculator.py --mode C --ram 289 --type gguf meta-llama/Meta-Llama-3.1-405B-Instruct-FP8
Q4_0
"""

File: language-model/gpu/vram_calculator.go
package main

// Calculate GPU VRAM requirements for LLM models
//
// Date created: 2024-08-11
// Version 1.8
// License: MIT
// Copyright: Cedric Chee

import (
	"encoding/json"
	"flag"
	"fmt"
	"io"
	"log"
	"math"
	"net/http"
	"os"
	"path/filepath"
	"sort"
	"strconv"
	"strings"
)

const (
	baseDir = "cache"
)

var (
	GGUF_MAPPING = map[string]float64{
		"Q8_0":    8.5,
		"Q6_K":    6.59,
		"Q5_K_M":  5.69,
		"Q5_K_S":  5.54,
		"Q5_0":    5.54,
		"Q4_K_M":  4.85,
		"Q4_K_S":  4.58,
		"Q4_0":    4.55,
		"IQ4_NL":  4.5,
		"Q3_K_L":  4.27,
		"IQ4_XS":  4.25,
		"Q3_K_M":  3.91,
		"IQ3_M":   3.7,
		"IQ3_S":   3.5,
		"Q3_K_S":  3.5,
		"Q2_K":    3.35,
		"IQ3_XS":  3.3,
		"IQ3_XXS": 3.06,
		"IQ2_M":   2.7,
		"IQ2_S":   2.5,
		"IQ2_XS":  2.31,
		"IQ2_XXS": 2.06,
		"IQ1_S":   1.56,
	}

	EXL2_OPTIONS = []float64{}
)

// bpwValue defines custom type that can hold either a float or a string value.
// The type is for the flag `-bpw“.
type bpwValue struct {
	value interface{}
}

// Set method attempts to parse the input string as a float, and if that fails,
// it stores the string value instead.
func (b *bpwValue) Set(s string) error {
	if f, err := strconv.ParseFloat(s, 64); err == nil {
		b.value = f
		return nil
	}
	b.value = s
	return nil
}

// String method returns a string representation of the value.
func (b *bpwValue) String() string {
	return fmt.Sprintf("%v", b.value)
}

func init() {
	i := 6.0
	for i > 1.0 {
		EXL2_OPTIONS = append(EXL2_OPTIONS, i)
		i -= 0.05
	}
}

func bitsToGb(bits float64) float64 {
	// return bits / (2 * 1024 * 1024 * 1024) // wrong
	return bits / math.Pow(2, 30)
}

func calculateVramRaw(
	numParams float64,
	bpw float64,
	lmHeadBpw float64,
	kvCacheBpw float64,
	context float64,
	fp8 bool,
	numGpus int,
	maxPositionEmbeddings float64,
	numHiddenLayers float64,
	hiddenSize float64,
	numKeyValueHeads float64,
	numAttentionHeads float64,
	intermediateSize float64,
	vocabSize float64,
	gqa bool,
) float64 {
	// cudaSize := 500 * 2 * 1024 * 1024 * float64(numGpus)
	// cudaSize := float64(500 * 1 << 20 * numGpus)
	cudaSize := float64(500) * math.Pow(2, 20) * float64(numGpus)
	paramsSize := numParams * 1e9 * (bpw / 8)
	kvCacheSize := (context * 2 * numHiddenLayers * hiddenSize) * (kvCacheBpw / 8)

	if gqa {
		kvCacheSize *= numKeyValueHeads / numAttentionHeads
	}

	bytesPerParam := bpw / 8
	lmHeadBytesPerParam := lmHeadBpw / 8

	headDim := hiddenSize / numAttentionHeads
	attentionInput := bytesPerParam * context * hiddenSize

	q := bytesPerParam * context * headDim * numAttentionHeads
	k := bytesPerParam * context * headDim * numKeyValueHeads
	v := bytesPerParam * context * headDim * numKeyValueHeads

	softmaxOutput := lmHeadBytesPerParam * numAttentionHeads * context
	softmaxDropoutMask := numAttentionHeads * context
	dropoutOutput := lmHeadBytesPerParam * numAttentionHeads * context

	outProjInput := lmHeadBytesPerParam * context * numAttentionHeads * headDim
	attentionDropout := context * hiddenSize

	attentionBlock := (attentionInput +
		q +
		k +
		softmaxOutput +
		v +
		outProjInput +
		softmaxDropoutMask +
		dropoutOutput +
		attentionDropout)

	mlpInput := bytesPerParam * context * hiddenSize
	activationInput := bytesPerParam * context * intermediateSize
	downProjInput := bytesPerParam * context * intermediateSize
	dropoutMask := context * hiddenSize
	mlpBlock := mlpInput + activationInput + downProjInput + dropoutMask

	layerNorms := bytesPerParam * context * hiddenSize * 2
	activationsSize := attentionBlock + mlpBlock + layerNorms

	outputSize := lmHeadBytesPerParam * context * vocabSize

	vramBits := cudaSize + paramsSize + activationsSize + outputSize + kvCacheSize

	return round(bitsToGb(vramBits), 2)
}

func downloadFile(url string, filename string, headers map[string]string) error {
	// Disable the following line for fixing error "stat cache/{model_id}/config.json: no such file or directory"
	// baseDir := filepath.Join(filepath.Dir(os.Args[0]), baseDir)
	filePath := filepath.Join(baseDir, filename)

	if _, err := os.Stat(filePath); err == nil {
		return nil
	}

	req, err := http.NewRequest("GET", url, nil)
	if err != nil {
		return err
	}

	for k, v := range headers {
		req.Header.Set(k, v)
	}

	resp, err := http.DefaultClient.Do(req)
	if err != nil {
		return err
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		return fmt.Errorf("status code: %d", resp.StatusCode)
	}

	os.MkdirAll(filepath.Dir(filePath), 0755)

	f, err := os.Create(filePath)
	if err != nil {
		return err
	}
	defer f.Close()

	_, err = io.Copy(f, resp.Body)
	if err != nil {
		return err
	}

	// Check if the file was saved correctly
	if _, err := os.Stat(filePath); err != nil {
		return err
	}

	return nil
}

func downloadModelConfigs(modelId string, access string) error {
	baseUrl := fmt.Sprintf("https://huggingface.co/%s/raw/main/", modelId)

	readmeUrl := baseUrl + "README.md"
	configUrl := baseUrl + "config.json"
	indexUrl := baseUrl + "model.safetensors.index.json"

	readmePath := filepath.Join(modelId, "README.md")
	configPath := filepath.Join(modelId, "config.json")
	indexPath := filepath.Join(modelId, "model.safetensors.index.json")

	headers := map[string]string{}
	if access != "" {
		headers["Authorization"] = fmt.Sprintf("Bearer %s", access)
	}

	if err := downloadFile(readmeUrl, readmePath, headers); err != nil {
		return err
	}
	if err := downloadFile(configUrl, configPath, headers); err != nil {
		return err
	}
	if err := downloadFile(indexUrl, indexPath, headers); err != nil {
		return err
	}

	return nil
}

func getModelConfig(modelId string, access string) (map[string]interface{}, error) {
	if err := downloadModelConfigs(modelId, access); err != nil {
		return nil, err
	}

	modelDir := filepath.Join(baseDir, modelId)
	configPath := filepath.Join(modelDir, "config.json")
	indexPath := filepath.Join(modelDir, "model.safetensors.index.json")

	// Check if the config file exists
	if _, err := os.Stat(configPath); err != nil {
		return nil, err
	}

	// Check if the index file exists
	if _, err := os.Stat(indexPath); err != nil {
		return nil, err
	}

	var config map[string]interface{}
	{
		f, err := os.Open(configPath)
		if err != nil {
			return nil, err
		}
		defer f.Close()

		if err := json.NewDecoder(f).Decode(&config); err != nil {
			return nil, err
		}
	}

	var index map[string]interface{}
	{
		f, err := os.Open(indexPath)
		if err != nil {
			return nil, err
		}
		defer f.Close()

		if err := json.NewDecoder(f).Decode(&index); err != nil {
			return nil, err
		}
	}

	numParams := index["metadata"].(map[string]interface{})["total_size"].(float64) / 2 / 1e9
	config["num_params"] = numParams
	return config, nil
}

func parseBpw(bpw string) float64 {
	f, err := strconv.ParseFloat(bpw, 64)
	if err != nil {
		return GGUF_MAPPING[strings.ToUpper(bpw)]
	}
	return f
}

func getBpwValues(bpw string, fp8 bool) map[string]float64 {
	bpwValue := parseBpw(bpw)

	lmHeadBpw := 6.0
	if bpwValue > 6.0 {
		lmHeadBpw = 8.0
	}

	kvCacheBpw := 16.0
	if fp8 {
		kvCacheBpw = 8.0
	}

	return map[string]float64{
		"bpw":          bpwValue,
		"lm_head_bpw":  lmHeadBpw,
		"kv_cache_bpw": kvCacheBpw,
	}
}

func calculateVram(modelId string, bpw string, context float64, fp8 bool, access string) (float64, error) {
	config, err := getModelConfig(modelId, access)
	if err != nil {
		return 0, err
	}

	bpwValues := getBpwValues(bpw, fp8)

	if context == 0 {
		context = config["max_position_embeddings"].(float64)
	}

	return calculateVramRaw(
		config["num_params"].(float64),
		bpwValues["bpw"],
		bpwValues["lm_head_bpw"],
		bpwValues["kv_cache_bpw"],
		context,
		fp8,
		1,
		config["max_position_embeddings"].(float64),
		config["num_hidden_layers"].(float64),
		config["hidden_size"].(float64),
		config["num_key_value_heads"].(float64),
		config["num_attention_heads"].(float64),
		config["intermediate_size"].(float64),
		config["vocab_size"].(float64),
		true,
	), nil
}

func calculateContext(modelId string, memory float64, bpw string, fp8 bool, access string) (float64, error) {
	config, err := getModelConfig(modelId, access)
	if err != nil {
		return 0, err
	}

	minContext := 2048.0
	maxContext := config["max_position_embeddings"].(float64)

	low, high := minContext, maxContext
	for low < high {
		mid := (low + high + 1) / 2
		vram, err := calculateVram(modelId, bpw, mid, fp8, access)
		if err != nil {
			return 0, err
		}
		if vram > memory {
			high = mid - 1
		} else {
			low = mid
		}
	}

	context := low
	for {
		vram, err := calculateVram(modelId, bpw, context, fp8, access)
		if err != nil {
			return 0, err
		}
		if vram >= memory || context >= maxContext {
			break
		}
		context += 100
	}

	return context - 100, nil
}

func calculateBpw(modelId string, memory float64, context float64, fp8 bool, type_ string, access string) (string, error) {
	if type_ == "exl2" {
		for _, bpw := range EXL2_OPTIONS {
			vram, err := calculateVram(modelId, fmt.Sprintf("%f", bpw), context, fp8, access)
			if err != nil {
				return "", err
			}
			if vram < memory {
				return fmt.Sprintf("%f", bpw), nil
			}
		}
	} else if type_ == "gguf" {
		// Be careful: The iteration order of Go maps won’t necessarily be the
		// same every time you run the program.

		// for name := range GGUF_MAPPING {
		// 	vram, err := calculateVram(modelId, name, context, fp8, access)
		// 	if err != nil {
		// 		return "", err
		// 	}
		// 	if vram < memory {
		// 		return name, nil
		// 	}
		// }

		// Ordering map keys
		keys := make([]string, 0, len(GGUF_MAPPING))
		for key := range GGUF_MAPPING {
			keys = append(keys, key)
		}
		sort.SliceStable(keys, func(i, j int) bool {
			return GGUF_MAPPING[keys[i]] > GGUF_MAPPING[keys[j]]
		})

		for _, k := range keys {
			name := k

			bpw := fmt.Sprintf("%f", GGUF_MAPPING[name])
			vram, err := calculateVram(modelId, bpw, context, fp8, access)
			if err != nil {
				return "", err
			}
			if vram < memory {
				return name, nil
			}
		}
	}
	return "", nil
}

func round(val float64, precision int) float64 {
	p := math.Pow10(precision)
	return math.Round(val*p) / p
}

func main() {
	flag.Usage = func() {
		w := flag.CommandLine.Output()
		fmt.Fprintf(w, "GPU VRAM, context length, and BPW calculator for LLMs.\n\n")
		fmt.Fprintf(w, "Usage: %s [options]\n\nOptions:\n", os.Args[0])
		flag.PrintDefaults()
		fmt.Fprintf(w, "\nMode-specific options:\n")
		fmt.Fprintf(w, "Mode A: Set -bpw and -context to get the amount of VRAM required\n")
		fmt.Fprintf(w, "Mode B: Set -memory and -bpw to get the amount of context you can fit in your VRAM\n")
		fmt.Fprintf(w, "Mode C: Set -memory and -context to get the best BPW you can fit in your VRAM\n\n")
		fmt.Fprintf(w, "Note:\n")
		fmt.Fprintf(w, "You can use numbers like 4.55 OR GGUF Quant IDs like Q4_0 for the -bpw value.\n")
		fmt.Fprintf(w, "For mode C, specifying -type gguf will return GGUF Quant IDs, while specifying -type exl2 will return floating point numbers.\n\n")
		fmt.Fprintf(w, "Examples:\n")
		fmt.Fprintf(w, "1) How much memory do I need to run a LLama 3.1 405B model quantized to Q4_0?\n")
		fmt.Fprintf(w, "./vram_calculator -model_id meta-llama/Meta-Llama-3.1-405B-Instruct-FP8 -mode A -bpw Q4_0 -access $HF_API_TOKEN\n\n")
		fmt.Fprintf(w, "2) How much context can I get out of this model?\n")
		fmt.Fprintf(w, "./vram_calculator -model_id meta-llama/Meta-Llama-3.1-405B-Instruct-FP8 -mode B -ram 289 -bpw 4.55 -access $HF_API_TOKEN\n\n")
		fmt.Fprintf(w, "3) What is the best quant I can run of this model?\n")
		fmt.Fprintf(w, "./vram_calculator -model_id meta-llama/Meta-Llama-3.1-405B-Instruct-FP8 -mode C -ram 289 -type gguf -access $HF_API_TOKEN\n\n")
	}

	modelId := flag.String("model_id", "", "Model ID from Hugging Face Hub")
	access := flag.String("access", "", "Access token for Hugging Face API (optional)")
	fp8 := flag.Bool("fp8", false, "Use FP8 KV cache (default false)")
	mode := flag.String("mode", "A", "Select mode")

	// The flag `-bpw` should accept either a float or a string value.
	bpw := &bpwValue{value: 5.0}
	// Define the `-bpw`` flag with the custom `bpwValue` type.
	flag.Var(bpw, "bpw", "Bits per weight (can be float or GGUF ID)")

	context := flag.Float64("context", 0, "Context length (default use model setting)")
	ram := flag.Float64("ram", 48, "Available VRAM in GB")
	type_ := flag.String("type", "exl2", "Type of quantization [exl2, gguf] for BPW calculation")

	flag.Parse()

	*mode = strings.ToLower(*mode)

	// Extract the float or string value from the `bpw` flag.
	bpwFloat, bpwStr := 0.0, ""
	switch v := bpw.value.(type) {
	case float64:
		bpwFloat = v
	case string:
		bpwStr = v
	default:
		log.Fatal("Invalid value for -bpw flag")
	}

	if *mode == "a" {
		// Mode A - calculate VRAM required

		vram, err := calculateVram(*modelId, bpwStr, *context, *fp8, *access)
		if bpwStr == "" {
			vram, err = calculateVram(*modelId, fmt.Sprintf("%f", bpwFloat), *context, *fp8, *access)
		}
		if err != nil {
			log.Fatal(err)
		}
		fmt.Println(vram)
	} else if *mode == "b" {
		// Mode B - calculate context you can fit in your VRAM

		context, err := calculateContext(*modelId, *ram, fmt.Sprintf("%f", bpwFloat), *fp8, *access)
		if bpwStr != "" {
			context, err = calculateContext(*modelId, *ram, bpwStr, *fp8, *access)
		}
		if err != nil {
			log.Fatal(err)
		}
		fmt.Println(int(context))
	} else if *mode == "c" {
		// Mode C - calculate the best BPW you can fit in your VRAM

		bpw, err := calculateBpw(*modelId, *ram, *context, *fp8, *type_, *access)
		if err != nil {
			log.Fatal(err)
		}
		fmt.Println(bpw)
	}
}

/*
Examples:

1) How much memory do I need to run a LLama 3.1 405B model quantized to Q4_0?
$ go run vram_calculator.go -model_id meta-llama/Meta-Llama-3.1-405B-Instruct-FP8 -mode A -bpw Q4_0 -access $HF_API_TOKEN
288.41


2) How much context can I get out of this model?
$ go run vram_calculator.go -model_id meta-llama/Meta-Llama-3.1-405B-Instruct-FP8 -mode B -ram 289 -bpw 4.55 -access $HF_API_TOKEN
130972


3) What is the best quant I can run of this model?
$ go run vram_calculator.go -model_id meta-llama/Meta-Llama-3.1-405B-Instruct-FP8 -mode C -ram 289 -type gguf -access $HF_API_TOKEN
Q4_0
*/
